taxi_demo/etl/ml_datasets.ipynb

```diff
@@ -0,0 +1,1985 @@
+       "   fare_amount  extra  mta_tax  tip_amount  tolls_amount  \\\n",
+       "0          8.9    0.5      NaN        0.00           0.0   \n",
+       "1         12.1    0.5      NaN        2.00           0.0   \n",
+       "2         23.7    0.0      NaN        4.74           0.0   \n",
+       "3         14.9    0.5      NaN        3.05           0.0   \n",
+       "4          3.7    0.0      NaN        0.00           0.0   \n",
+       "\n",
+       "   improvement_surcharge  total_amount  congestion_surcharge  \n",
+       "0                    NaN          9.40                   NaN  \n",
+       "1                    NaN         14.60                   NaN  \n",
+       "2                    NaN         28.44                   NaN  \n",
+       "3                    NaN         18.45                   NaN  \n",
+       "4                    NaN          3.70                   NaN  \n",
+       "\n",
+       "[5 rows x 22 columns]"
+      ]
+     },
+     "execution_count": 3,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "taxi = dd.read_parquet(f'{taxi_path}/data/taxi_parquet', engine='pyarrow')\n",
+    "taxi.head()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 73,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CPU times: user 190 ms, sys: 11.8 ms, total: 202 ms\n",
+      "Wall time: 11.4 s\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "1611604226"
+      ]
+     },
+     "execution_count": 73,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "%%time\n",
+    "len(taxi)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# only use zones for ML tasks\n",
+    "# change to spatial join to pull in old data if there's time\n",
+    "taxi = taxi[taxi.pickup_taxizone_id.notnull() & taxi.dropoff_taxizone_id.notnull()]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create some features and split data\n",
+    "\n",
+    "- These features are stateless, so we can compute them before splitting into train/test sets\n",
+    "    - Can you think of more features that might make our models better?\n",
+    "- Use last 6 months of 2019 as test set, all other data for train set"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "taxi['tip_fraction'] = taxi.tip_amount / taxi.total_amount\n",
+    "taxi['pickup_weekday'] = taxi.pickup_datetime.dt.weekday\n",
+    "taxi['pickup_weekofyear'] = taxi.pickup_datetime.dt.weekofyear\n",
+    "taxi['pickup_hour'] = taxi.pickup_datetime.dt.hour\n",
+    "taxi['pickup_minute'] = taxi.pickup_datetime.dt.minute\n",
+    "taxi['pickup_week_hour'] = (taxi.pickup_weekday * 24) + taxi.pickup_hour"
```

Comment:
I think it's important to add text in this notebook explaining why we chose only stateless features. That is the type of thing that will:

1. make the code easier for a user to understand
2. show that we know what we're doing and made these choices thoughtfully

You put a lot of careful thought into which features we do and don't want, because of the deployment concerns associated with stateful stuff or scaling, so we should share those thoughts here!